test/utils/floating_point_utils.hpp

```diff
@@ -0,0 +1,81 @@
+// Fizzy: A fast WebAssembly interpreter
+// Copyright 2020 The Fizzy Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <cmath>
+#include <cstdint>
+#include <limits>
+#include <type_traits>
+
+namespace fizzy::test
+{
+/// Simple implementation of C++20's std::bit_cast.
+template <typename DstT, typename SrcT>
+DstT bit_cast(SrcT x) noexcept
+{
+    DstT z;
+    static_assert(sizeof(x) == sizeof(z));
+    __builtin_memcpy(&z, &x, sizeof(x));
+    return z;
+}
+
+/// A wrapper for floating-point types with inspection/construction/comparison utilities.
+template <typename T>
+struct FP
+{
+    static_assert(std::is_floating_point_v<T>);
+
+    /// Shortcut to numeric_limits.
+    using limits = std::numeric_limits<T>;
+
+    /// The unsigned integer type matching the size of this floating-point type.
+    using uint_type = std::conditional_t<std::is_same_v<T, float>, uint32_t, uint64_t>;
+    static_assert(sizeof(T) == sizeof(uint_type));
+
+    /// The number of mantissa bits in the binary representation.
+    static constexpr auto num_mantissa_bits = limits::digits - 1;
+
+    /// The binary mask of the mantissa part of the binary representation.
+    static constexpr auto mantissa_mask = (uint_type{1} << num_mantissa_bits) - 1;
+
+    /// The payload of the canonical NaN (only the top bit set).
+    /// See: https://webassembly.github.io/spec/core/syntax/values.html#canonical-nan.
+    static constexpr auto canon = uint_type{1} << (num_mantissa_bits - 1);
+
+    T value{};
+
+    explicit FP(T v) noexcept : value{v} {};
+
+    explicit FP(uint_type u) noexcept : value{bit_cast<T>(u)} {};
+
+    /// Return unsigned integer with the binary representation of the value.
+    uint_type as_uint() const noexcept { return bit_cast<uint_type>(value); }
+
+    /// Returns NaN payload if the value is a NaN, otherwise 0 (NaN payload is never 0).
+    uint_type nan_payload() const noexcept
+    {
+        return std::isnan(value) ? (as_uint() & mantissa_mask) : 0;
+    }
+
+    /// Build the NaN value with the given payload.
+    static T nan(uint_type payload = canon) noexcept
+    {
+        const auto nan_payload = payload & mantissa_mask;
+        const auto nan_exponent = (uint_type{limits::max_exponent} << 1) - 1;
```

Comment:
This line is a bit confusing to me...
Why `<< 1`?

Reply:
The exponent is kinda signed value, e.g. for `float32` exponent has 8 bits with values range `[-127, 128]`. To get exponent with all bits set you need something like `2 * max_exponent - 1`. Can change to multiplication if that's better.